### PR TITLE
Abbreviations inside formatters and other commands

### PR DESCRIPTION
**Pull request by JelleAalbers**
This allows reverse abbreviations inside formatters and other commands:
  * _funky read brief text_ → def read_txt():
  * _all caps snake brief command something_ → CMD_SOMETHING
  * _say fancy brief technology comma cool things brief etcetera_ → fancy tech, cool things etc

The implementation is to add several _brief SPOKEN_FORM_  → ABBREVIATION entries to talon's `user.vocabulary` map in a command-mode context. 

Some notes:
  * I'm too new to Talon to know how appropriate / evil messing with `user.vocabulary` is. Maybe there are better solutions? I tried using `dictate.word_map`, but that does not support phrases, as noted in #389.
  * Comments in `code/vocabulary.py` indicate that `user.vocabulary` phrases are given high recognition priority. Since the added phrases all have multiple words starting with 'brief', I imagine this is fine?
  * Surprises can happen if the literal word 'brief' is intended. For example, _snake brief phrase_ → brief_phrase is fine, but _snake brief command_ → cmd, and  _say please make a brief entry_ → please make a ent
  * I left dictation mode alone here, but we could instead update `user.vocabulary` in the global context. That would help dictate things like the third example at the top of the PR.
  * Would it be worth it to also support the long version of the reverse-abbreviate command (_abbreviate X_) inside other commands?

### CODE REVIEW

**Comment by rntz** [comment]
I think this is a useful feature, although I would like input from other maintainers on it.

However, I don't think this is the right way to implement it. It involves a cross-python-script import, which we generally try to avoid because it reduces resilience to errors in user scripts. Also, it deeply breaks the abstraction boundary of vocabulary.py - someone editing vocabulary.py would have to avoid renaming `additional_words` or otherwise changing up how vocabulary.py edits the word list.

A better way would be to roll this feature into dictation.py (see below) by editing the captures that define `<user.word>`, `<user.text>`, and `<user.prose>` to include a `brief {user.abbreviation}` branch.

https://github.com/knausj85/knausj_talon/blob/3391fc9c6af2c86f3951b9edf0687880848b9b47/code/dictation.py#L23

**Comment by pokey** [comment]
Worth testing what this one does to DFA too. I added a bunch of this kind of stuff to my formatters and then removed it because my DFA exploded

**Comment by JelleAalbers** [comment]
Thanks for the quick feedback!

Great idea @rntz. I just tried adding a new `{user.inline_abbreviation}` option (containing the abbreviations with 'brief' added to their spoken forms) to the captures you mentioned; this works fine and avoids the relative import. If I add `brief {user.abbreviation}` it would also capture the word 'brief', but maybe there is some anti-capturing syntax to avoid that? Just let me know which version you and other maintainers would prefer.

Thanks for the heads-up @pokey. I followed your instructions [here](https://talonvoice.slack.com/archives/CCEE45RT7/p1635427836108300?thread_ts=1635402432.099900&cid=CCEE45RT7) to measure the compile time on issuing a first command when switching to a new application. (I don't always see this triggering compilation though -- maybe Talon does some caching?)

|           | gedit | vscode | vscode + cursorless |
| --- | --- | --- | --- |
| Master branch | 72 ms | 118 ms | 361 ms |
| This PR, extending `user.vocabulary` | 147 ms | 318 ms | 644 ms |
| This PR, `inline_abbreviation` in `dictation.py` captures | 187 ms | 405 ms | 687 ms |

Interestingly it seems to impact richer contexts more, and extending `user.vocabulary` is slightly cheaper than modifying the three dictation.py captures.

I don't think there is much/any extra delay for subsequent commands, though I'm not sure how to benchmark that accurately.

**Comment by rntz** [comment]
Yeah, given those DFA times I think we probably shouldn't ship this, unfortunately :(. 644ms is pretty bad. There are some changes upcoming in Talon that should reduce the impact large lists have on dfa compilation time, hopefully we can merge this after those land, but no guarantees when that will happen.

**Comment by rntz** [comment]
Btw, the approach I'd take to doing this is to use an intermediate capture rather than a list. That avoids creating another large list just to add "brief" to the start of things. Something like:

```
@mod.capture(rule="({user.vocabulary} | <user.abbreviation> | <phrase>)+")
def text(m) -> str:
    """A sequence of words, including user-defined vocabulary."""
    return format_phrase(m)
    
@mod.capture(rule="brief {user.abbreviation}")
def abbreviation(m) -> str:
    """DOCSTRING HERE"""
    return m.abbreviation
```

I haven't tested this but I think it should work.

**Comment by JelleAalbers** [comment]
Sure, I understand. Your version works great and is certainly cleanest. I just pushed that for future reference; feel free to close this or leave it open for later.

**Comment by splondike** [comment]
We'll leave this open until we get a change in DFA times and can then revisit. Tagging and leaving open.

**Comment by rntz** [comment]
DFA times have improved in beta but we're now waiting on those changes being available in public talon to merge this.

**Comment by rntz** [comment]
Hopefully those improvements will go out in 0.3 soonish but we might still want to somehow make this conditional on using w2l instead of dragon, since dragon can't handle large lists.

**Comment by rntz** [comment]
Ok, the DFA improvements are now in public. So the only thing holding this up is the question of not wrecking Dragon users. Ideally we can gate that on `ctx.matches = "speech.engine: wav2letter"` somehow.

File: code/dictation.py
Comment (rntz) [inline code/dictation.py:38]: We should test whether Dragon performs OK using this definition. If it doesn't, we should make this a contextual override for the capture instead of its global definition, in a context matching `speech.engine: wav2letter`.

**Review by rntz — CHANGES_REQUESTED** [review]

File: code/dictation.py
Comment (nriley) [reply code/dictation.py:38]: @wolfmanstout are you still using Dragon and would you be able to take a look?

Comment (wolfmanstout) [reply code/dictation.py:38]: I'm not using Dragon any more regularly. Last I checked it was already really slow. If no one else has Dragon I can try to find some time to test, but I don't have much time these days so it may not be this weekend.

Comment (pokey) [reply code/dictation.py:38]: Hmm do we know anyone else using Dragon these days?  Maybe we should ping on slack to see if anyone is willing to help out and test this branch?

**Comment by splondike** [comment]
phillco's going to push this change along some by testing it with Dragon to confirm that it can't handle it.

If indeed it can't then we can look at conditionally implementing the capture based on the `speech.engine: wav2letter` matcher rntz identified; you can have different rules for captures in different contexts (see the wiki [captures vs lists in voice commands](https://talon.wiki/unofficial_talon_docs/#captures-vs-lists-in-voice-commands) section).

File: code/dictation.py
Comment (splondike) [inline code/dictation.py:15]: Mind elaborating what advantage there is to having <user.abbreviated> in the <user.word> capture? My read is this change would just let you say 'word brief difference' and get 'diff' output. You could also just say 'brief difference' and get the same thing.

**Comment by splondike** [comment]
@phillco This ticket hasn't really moved in almost a year. I'm thinking we just close it out since nobody seems to be pushing it along. What do you reckon?

**Comment by rntz** [comment]
I would like to see this merged, actually. I think it just dropped off the radar, but it seems useful.

**Comment by rntz** [comment]
If nobody can try it with Dragon soon, then I think we should just assume Dragon can't handle it on the basis of prior experience and use the speech.engine approach.

File: code/dictation.py
Comment (fidgetingbits) [reply code/dictation.py:15]: I think the reasoning here is that if there are commands that take `<word>` as an argument, like for instance `proud <word>`, this allows you to use the abbreviated version as the `<word>`. It seems in that case it's worth keeping it.

**Comment by knausj85** [comment]
This is out of date enough that I think it's easier to close and reimplement, which I'll do now.

**Pull request closed without merging**
